src/nav2_tasks/include/nav2_tasks/task_client.hpp

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_TASKS__TASK_CLIENT_HPP_
+#define NAV2_TASKS__TASK_CLIENT_HPP_
+
+#include <atomic>
+#include <condition_variable>
+#include <string>
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+#include "nav2_tasks/task_status.hpp"
+
+namespace nav2_tasks
+{
+
+template<class CommandMsg, class ResultMsg>
+class TaskClient
+{
+public:
+  TaskClient(const std::string & name, rclcpp::Node * node)
+  : node_(node)
+  {
+    // Create the publishers
+    commandPub_ = node_->create_publisher<CommandMsg>(name + "_command");
+    cancelPub_ = node_->create_publisher<CancelMsg>(name + "_cancel");
+
+    // Create the subscribers
+    resultSub_ = node_->create_subscription<ResultMsg>(name + "_result",
+        std::bind(&TaskClient::onResultReceived, this, std::placeholders::_1));
+    statusSub_ = node_->create_subscription<std_msgs::msg::String>(name + "_status",
+        std::bind(&TaskClient::onStatusReceived, this, std::placeholders::_1));
+  }
+
+  ~TaskClient()
+  {
+  }
+
+  // The client can tell the TaskServer to execute its operation
+  void executeAsync(const typename CommandMsg::SharedPtr msg)
```

Comment:
I think this would be better named "sendAsyncCommand", or as they do in rclcpp services "send_async_request". executeAsync is a bit confusing (at least to me) as the execution is actually on the server side. We're just requesting/commanding that execution to begin.